src/dataflow/state.rs

```diff
@@ -1,25 +1,43 @@
 //! Structures and traits for states added to streams.
 
+use crate::dataflow::Timestamp;
+
 // TODO: keep around messages. Add an iterator over messages.
 // Add set_timestamp and set_access_context to State.
-use std::{
-    collections::BTreeMap,
-    ops::Bound::{Excluded, Unbounded},
-};
-
-use crate::dataflow::Timestamp;
 
 /// Trait that must be implemented by stream state.
 pub trait State: 'static + Clone + Send + Sync {}
 impl<T: 'static + Clone + Send + Sync> State for T {}
 
+
+// TODO (Sukrit): Do these state traits also require a way to read the state for a given timestamp?
+
+/// Trait that must be implemented by a state structure that is used in a Writeable operator.
+/// This structure must implement a `commit` method that commits the final state for a given
```

Comment:
Should `commit` only be accessible from within a watermark callback? Could also automatically commit after a watermark is executed.

Also, why is `commit` implemented by the user? Shouldn't the system handle time-versioned state?

Reply:
The API does not prevent anyone from committing inside a message callback right now since both `on_data` and `on_watermark` receive a mutable reference to `WriteableSinkContext`. We can choose to automatically commit, but I guess it would be better if the developer can commit state and continue some cleanup / message sending work.

I envision only power users implementing `commit` (or even `append` in `ReadOnlyState`). Ideally, the system would provide a default State implementation that implements `ReadOnlyState` and `WriteableState` traits, which most users would use.

The current PR does not have a system-defined implementation for these two traits.

Reply:
If that's the case, shouldn't state commits happen automatically from watermark callbacks? I'm leaning towards keeping a simpler, smaller interface since we can always expose `commit` later when a use case arises.

Also, if `commit` and `append` are intended for power-users, we should provide default implementations, e.g.
`fn commit(&self, state: &T, timestamp: &Timestamp) {};`

Reply:
Fixed in the new state implementation and the watermark callbacks now automatically commit state upon completion.